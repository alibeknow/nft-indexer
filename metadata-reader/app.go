package metadata_reader

import (
	"context"
	"time"

	"github.com/prometheus/client_golang/prometheus"
	"github.com/zeebo/errs"
	"go.uber.org/zap"
	"golang.org/x/sync/errgroup"

	"metadata-reader/database"
	"metadata-reader/private/httpreader"
	"metadata-reader/queue"
	"metadata-reader/server"
)

var (
	requestPerSecond = prometheus.NewGauge(
		prometheus.GaugeOpts{
			Name: "request_per_second_gauge",
			Help: "Number of requests per second",
		},
	)
	messagePerSecond = prometheus.NewGauge(
		prometheus.GaugeOpts{
			Name: "message_per_second_gauge",
			Help: "Number of messages per second",
		},
	)
	instantRequestPerSecond = prometheus.NewGauge(
		prometheus.GaugeOpts{
			Name: "instant_request_per_second_gauge",
			Help: "Instant number of reqs per second",
		},
	)

	instantMessagePerSecond = prometheus.NewGauge(
		prometheus.GaugeOpts{
			Name: "instant_messages_per_second_gauge",
			Help: "Instant number of reqs per second",
		},
	)
)

func init() {
	prometheus.MustRegister(requestPerSecond)
	prometheus.MustRegister(messagePerSecond)
	prometheus.MustRegister(instantRequestPerSecond)
	prometheus.MustRegister(instantMessagePerSecond)
}

type App struct {
	log      *zap.Logger
	db       *database.DB
	srv      *server.Server
	listener *queue.Listener
}

func NewApp(log *zap.Logger, db *database.DB, srv *server.Server, listener *queue.Listener) (*App, error) {
	app := &App{
		log:      log,
		db:       db,
		srv:      srv,
		listener: listener,
	}

	return app, nil
}

func (mr *App) Run(ctx context.Context) error {
	group, ctx := errgroup.WithContext(ctx)

	group.Go(func() error {
		now := time.Now()
		ticker := time.NewTicker(time.Second)
		defer ticker.Stop()

		var (
			reqLast int64
			msgLast int64
		)

		for {
			select {
			case <-ctx.Done():
				return nil
			case <-ticker.C:
				duration := time.Since(now).Seconds()

				reqCount := httpreader.TotalRequestCount()
				requestPerSecond.Set(float64(reqCount) / duration)
				reqDelta := reqCount - reqLast
				reqLast = reqCount
				instantRequestPerSecond.Set(float64(reqDelta))

				msgCount := queue.TotalMessageCount()
				messagePerSecond.Set(float64(msgCount) / duration)
				msgDelta := msgCount - msgLast
				msgLast = msgCount
				instantMessagePerSecond.Set(float64(msgDelta))
			}
		}
	})
	group.Go(func() error {
		return mr.srv.Run(ctx)
	})
	group.Go(func() error {
		return mr.listener.Run(ctx)
	})

	return group.Wait()
}

func (mr *App) Close() error {
	serverErr := mr.srv.Close()
	listenerErr := mr.listener.Close()

	return errs.Combine(serverErr, listenerErr)
}
