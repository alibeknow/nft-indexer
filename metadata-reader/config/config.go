package config

import (
	"bytes"
	"encoding/json"
	"fmt"
	"log"
	"os"
	"strings"

	"github.com/spf13/viper"

	"metadata-reader/database"
	"metadata-reader/metadata"
	"metadata-reader/private/logutils"
	"metadata-reader/private/opensearch"
	"metadata-reader/queue"
	"metadata-reader/server"
)

type Config struct {
	Logger   logutils.Config
	Server   server.Config
	Database database.Config
	Queue    queue.Config
	Metadata metadata.Config
	Search   opensearch.Config
}

func LoadConfig(configPaths ...string) (Config, error) {
	v := viper.New()
	v.SetConfigName("config")
	v.SetConfigType("yaml")
	v.SetEnvPrefix("MR")
	v.SetEnvKeyReplacer(strings.NewReplacer(`.`, `_`))
	v.AutomaticEnv()

	v2 := viper.New()
	v2.SetConfigName("config")
	v2.SetConfigType("yaml")
	var buff bytes.Buffer
	m := map[string]interface{}{}
	err := json.NewEncoder(&buff).Encode(Config{})
	if err != nil {
		return Config{}, fmt.Errorf("failed to encode config: %v", err)
	}
	err = json.NewDecoder(&buff).Decode(&m)
	if err != nil {
		return Config{}, fmt.Errorf("failed to decode config: %v", err)
	}
	for key, val := range m {
		v2.Set(key, val)
	}

	for _, path := range configPaths {
		v.AddConfigPath(path)
		v2.AddConfigPath(path)
	}

	if err = v2.SafeWriteConfig(); err != nil {
		if _, ok := err.(viper.ConfigFileAlreadyExistsError); ok {
			log.Println("config.yaml file already exists: skip creating")
		} else {
			return Config{}, fmt.Errorf("failed to write the configuration file: %v", err)
		}
	}

	if err = v.ReadInConfig(); err != nil {
		if _, ok := err.(viper.ConfigFileNotFoundError); ok {
			log.Println("config.yaml ignored")
		} else {
			return Config{}, fmt.Errorf("failed to read the configuration file: %v", err)
		}
	}

	for _, k := range v.AllKeys() {
		value := v.GetString(k)
		v.Set(k, os.ExpandEnv(value))
	}

	var config Config
	err = v.Unmarshal(&config)
	return config, err
}
