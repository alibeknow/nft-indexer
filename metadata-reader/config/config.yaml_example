server:
  address: ":xxx"

database:
  name: "xxx"
  protocol: "mongodb"
  username: "xxx"
  password: "xxx"
  host: "xxx" #should be "localhost" for local use, and "mongodb" for docker
  port: xxx #number
  sslEnable: true/false
  directConnection: true/false

queue:
  protocol: "amqp"
  username: "xxx"
  password: "xxx"
  host: "xxx" #should be "localhost" for local use, and "rabbitmq" for docker
  port: xxx #number
  name: "xxx"

reader:
  workers: xxx #number
  proxyUri: "xxx"
  ipfsGateway: "https://ipfs.io/ipfs/"
  pinataDefaultGateway: "https://gateway.pinata.cloud/ipfs/"

httpreader:
  retries: xxx #number
