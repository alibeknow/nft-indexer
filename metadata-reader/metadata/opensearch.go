package metadata

import (
	"encoding/json"
	"fmt"
)

type MetadataDocument struct {
	Name            string                 `json:"name"`
	Description     string                 `json:"description"`
	ContractAddress string                 `json:"contractAddress"`
	Attributes      []map[string]string    `json:"attributes"`
	Chain           Chain                  `json:"chain"`
	Metadata        map[string]interface{} `json:"metadata"`
}

func MapToMetadataDocument(contractAddress string, chain Chain, rawMetadata string) (MetadataDocument, error) {
	var metadata map[string]interface{}
	var props struct {
		Attributes []map[string]interface{} `json:"attributes"`
		Properties map[string]interface{}   `json:"properties"`
	}

	err := json.Unmarshal([]byte(rawMetadata), &metadata)
	if err != nil {
		return MetadataDocument{}, err
	}

	_ = json.Unmarshal([]byte(rawMetadata), &props)

	attributes := []map[string]string{}
	for _, attribute := range props.Attributes {
		var key interface{}
		if value, found := attribute["trait_type"]; found {
			key = value
		}

		attributes = append(attributes, map[string]string{
			"key":   fmt.Sprint(key),
			"value": fmt.Sprint(attribute["value"]),
		})
	}

	for key, value := range props.Properties {
		b, err := json.Marshal(value)
		if err != nil {
			return MetadataDocument{}, err
		}

		attributes = append(attributes, map[string]string{
			"key":   key,
			"value": string(b),
		})
	}

	name, _ := metadata["name"].(string)
	description, _ := metadata["description"].(string)

	return MetadataDocument{
		Name:            name,
		Description:     description,
		ContractAddress: contractAddress,
		Attributes:      attributes,
		Chain:           chain,
		Metadata:        metadata,
	}, nil
}
