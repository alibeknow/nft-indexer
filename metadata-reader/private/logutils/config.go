package logutils

type Config struct {
	Level string
}
