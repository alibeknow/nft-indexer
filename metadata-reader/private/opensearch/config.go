package opensearch

type Config struct {
	Uri      string
	Shards   int
	Replicas int
}
