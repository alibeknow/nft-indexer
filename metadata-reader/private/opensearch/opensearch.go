package opensearch

import (
	"context"
	"crypto/tls"
	"encoding/json"
	"errors"
	"io"
	"net/http"
	"strings"

	opensearch "github.com/opensearch-project/opensearch-go"
	opensearchapi "github.com/opensearch-project/opensearch-go/opensearchapi"
	"go.uber.org/zap"
)

type OpenSearch struct {
	log      *zap.Logger
	client   opensearch.Client
	shards   int
	replicas int
}

func New(log *zap.Logger, config Config) (*OpenSearch, error) {
	client, err := opensearch.NewClient(opensearch.Config{
		Transport: &http.Transport{
			TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
		},
		Addresses: []string{config.Uri},
	})
	if err != nil {
		return nil, err
	}

	return &OpenSearch{
		client:   *client,
		log:      log,
		shards:   config.Shards,
		replicas: config.Replicas,
	}, nil
}

func (opensearch *OpenSearch) mapMappingsToSettings(mappings map[string]interface{}) (io.Reader, error) {
	var settings struct {
		Settings struct {
			Index struct {
				NumberOfShards   int `json:"number_of_shards"`
				NumberOfReplicas int `json:"number_of_replicas"`
			} `json:"index"`
		} `json:"settings"`
		Mappings map[string]interface{} `json:"mappings"`
	}

	settings.Settings.Index.NumberOfShards = opensearch.shards
	settings.Settings.Index.NumberOfReplicas = opensearch.replicas
	settings.Mappings = mappings

	result, err := json.Marshal(settings)
	if err != nil {
		return nil, err
	}

	return strings.NewReader(string(result)), err
}

func (opensearch *OpenSearch) InitIndex(ctx context.Context, indexName string, mappings map[string]interface{}) error {
	exists, err := opensearch.IndexExists(ctx, indexName)
	if err != nil {
		return err
	}
	if exists {
		return nil
	}

	opensearch.log.Info("Opensearch index does not exist: creating", zap.String("index", indexName))
	err = opensearch.CreateIndex(ctx, indexName, mappings)
	if err != nil {
		return err
	}

	opensearch.log.Info("Opensearch index created", zap.String("index", indexName))

	return nil
}

func (opensearch *OpenSearch) CreateIndex(ctx context.Context, indexName string, mappings map[string]interface{}) error {
	settings, err := opensearch.mapMappingsToSettings(mappings)
	if err != nil {
		return err
	}

	req := opensearchapi.IndicesCreateRequest{
		Index: indexName,
		Body:  settings,
	}

	resp, err := req.Do(ctx, &opensearch.client)
	if err != nil {
		return err
	}
	if resp.IsError() {
		return errors.New(resp.String())
	}

	return nil
}

func (opensearch *OpenSearch) IndexExists(ctx context.Context, indexName string) (bool, error) {
	existsReq := opensearchapi.IndicesExistsRequest{
		Index: []string{indexName},
	}

	resp, err := existsReq.Do(ctx, &opensearch.client)
	if err != nil {
		return false, err
	}

	return !resp.IsError(), nil
}

func (opensearch *OpenSearch) UpsertToIndex(ctx context.Context, indexName string, docId string, body io.Reader) (*opensearchapi.Response, error) {
	request := opensearchapi.IndexRequest{
		Index:      indexName,
		DocumentID: docId,
		Body:       body,
	}

	resp, err := request.Do(ctx, &opensearch.client)
	if err != nil {
		return nil, err
	}
	if resp.IsError() {
		return resp, errors.New(resp.String())
	}

	return resp, nil
}
