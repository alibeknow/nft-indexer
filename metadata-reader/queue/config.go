package queue

import "fmt"

type Config struct {
	Protocol string
	Username string
	Password string
	Host     string
	Port     int
	Name     string
	Workers  int
}

func (config *Config) URL() string {
	url := fmt.Sprintf(
		"%s://%s:%s@%s:%d",
		config.Protocol,
		config.Username,
		config.Password,
		config.Host,
		config.Port,
	)

	return url
}
