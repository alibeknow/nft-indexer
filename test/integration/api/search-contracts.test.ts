import request from 'supertest';
import nock from 'nock';
import { getApiConfig } from '@api/app.config';
import { HttpServer, INestApplication } from '@nestjs/common';
import { ConfigModule } from '@nestjs/config';
import { Test } from '@nestjs/testing';
import { IBaseConfig } from '@shared/baseconfig';
import { OpensearchModule, OpensearchProvider } from '@shared/opensearch';
import { LoggerModule } from 'nestjs-pino';
import { IContractIndex, mapToSearchContractIndex } from '@shared/contracts/contract-search';
import { getNestApp } from '../../helpers/app';
import { SearchModule } from '@api/search';
import { fixtureContractsDB } from './fixtures/contracts-db.fixture';
import { fixtureTokensDB } from './fixtures/tokens-db.fixture';
import { openseaResponseFixture } from './fixtures/opensea-response.fixture';

const collection = openseaResponseFixture.collection;

nock('https://test.opensea/').persist().get(/.*/).reply(200, { collection });

function seedContractsOpenSearch(opensearch: OpensearchProvider<IBaseConfig>, indexname: string) {
  return Promise.all(fixtureContractsDB.map( async (contract) => {
    const modifiedData = await mapToSearchContractIndex(contract, fixtureTokensDB.length);

    return opensearch
      .addDocumentToIndex(indexname, contract._id, modifiedData as IContractIndex);
  }) );
}

jest.setTimeout(20000);

describe('[Contracts search endpoint]', () => {
  let app: INestApplication;
  let httpServer: HttpServer;
  let opensearchService: OpensearchProvider<IBaseConfig>;
  const indexName = 'contractstest';

  beforeAll(async () => {
    const moduleRef = await Test.createTestingModule({
      imports: [
        ConfigModule.forRoot({
          load: [ getApiConfig ],
          isGlobal: true,
        }),
        OpensearchModule,
        SearchModule,
        LoggerModule.forRoot(),
      ],
    }).compile();
    opensearchService = moduleRef.get<OpensearchProvider<IBaseConfig>>(OpensearchProvider<IBaseConfig>);
    app = await getNestApp(moduleRef);
    httpServer = app.getHttpServer();
  });

  beforeEach(async () => {
    await seedContractsOpenSearch(opensearchService, indexName);
  });

  describe('[GET /api/v0/search/contracts]', () => {
    it('should return 200', async () => {
      const actualResult = await request(httpServer)
        .get('/api/v0/search/contracts?q=Art Blocks&filter=name&chain=eth')
        .set('Accept', 'application/json');

      expect(actualResult.body).toMatchSnapshot();
      expect(actualResult.statusCode).toBe(200);
      expect(actualResult.body.result.total).toBe(2);
    });
  });

  describe('[GET /api/v0/search/contracts/address]', () => {
    it('should return 200', async () => {
      const actualResult = await request(httpServer)
        .get('/api/v0/search/contracts/0x9FB2EEb75754815c5Cc9092Cd53549cEa5dc404f?q=Art Blocks&filter=name&chain=eth')
        .set('Accept', 'application/json');

      expect(actualResult.body).toMatchSnapshot();
      expect(actualResult.statusCode).toBe(200);
      expect(actualResult.body.result.total).toBe(1);
    });
  });

  afterAll(async () => {
    await opensearchService.removeIndex(indexName);
    await app.close();
  });
});
