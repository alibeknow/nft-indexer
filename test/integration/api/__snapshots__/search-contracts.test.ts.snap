// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Contracts search endpoint] [GET /api/v0/search/contracts/address] should return 200 1`] = `
{
  "result": {
    "from": 0,
    "hits": [
      {
        "collection_image": "123",
        "contract_address": "0x9FB2EEb75754815c5Cc9092Cd53549cEa5dc404f",
        "contract_description": "",
        "contract_name": "Superlative Mutated Apes",
        "contract_type": "ERC-721",
        "created_at": null,
        "number_of_tokens": 4,
        "synced_at": null,
        "verified": false,
      },
    ],
    "size": 15,
    "total": 1,
  },
  "statusCode": 200,
}
`;

exports[`[Contracts search endpoint] [GET /api/v0/search/contracts] should return 200 1`] = `
{
  "result": {
    "from": 0,
    "hits": [
      {
        "collection_image": "some image",
        "contract_address": "0xa7d8d9ef8D8Ce8992Df33D8b8CF4Aebabd5bD270",
        "contract_description": "",
        "contract_name": "Art Blocks",
        "contract_type": "ERC-721",
        "created_at": null,
        "number_of_tokens": 4,
        "synced_at": null,
        "verified": false,
      },
      {
        "collection_image": "123",
        "contract_address": "0xa7d8d9ef8D8Ce8992Df33D8b8CF4Aebabd5bD271",
        "contract_description": "",
        "contract_name": "Art Blocks",
        "contract_type": "ERC-721",
        "created_at": null,
        "number_of_tokens": 4,
        "synced_at": null,
        "verified": false,
      },
    ],
    "size": 15,
    "total": 2,
  },
  "statusCode": 200,
}
`;
